### PR TITLE
SONAR-6834 filter components by name in api/ce/activity

### PR DESCRIPTION
**Pull request by simonbrandhof**
SONAR-6834 filter components by name in api/ce/activity

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java [hunk 1]
```diff
@@ -36,15 +38,22 @@
...
 import org.sonar.server.ws.WsUtils;
 import org.sonarqube.ws.Common;
 import org.sonarqube.ws.WsCe;
 
+import static java.lang.String.format;
+
 public class ActivityWsAction implements CeWsAction {
 
   private static final String PARAM_COMPONENT_UUID = "componentId";
+  private static final String PARAM_COMPONENT_QUERY = "componentQuery";
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: please use the generic way to handle query: an example https://github.com/SonarSource/sonarqube/blob/adb1bccb465775981bb6b6c85c1df4d3ad9191dd/server/sonar-server/src/main/java/org/sonar/server/permission/ws/SearchProjectPermissionsAction.java#L69-69

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java [hunk 2]
```diff
@@ -93,7 +105,9 @@ public void define(WebService.NewController controller) {
   public void handle(Request wsRequest, Response wsResponse) throws Exception {
     DbSession dbSession = dbClient.openSession(false);
     try {
-      CeActivityQuery query = readQuery(wsRequest);
+      CeActivityQuery query = buildQuery(dbSession, wsRequest);
+      verifyPermissions(query);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: in general I use checkPermissions as a naming convention. I'd like us to be consistent

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java [hunk 3]
```diff
@@ -111,7 +125,25 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
...
+    } else {
+      userSession.checkGlobalPermission(UserRole.ADMIN);
+    }
+  }
+
+  private CeActivityQuery buildQuery(DbSession dbSession, Request wsRequest) {
+    String componentUuid = wsRequest.param(PARAM_COMPONENT_UUID);
+    String componentQuery = wsRequest.param(PARAM_COMPONENT_QUERY);
+    if (componentUuid != null && componentQuery != null) {
+      throw new BadRequestException("");
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: add a more meaningful message

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java [hunk 4]
```diff
@@ -123,17 +155,25 @@ private CeActivityQuery readQuery(Request wsRequest) {
...
-    if (componentUuid == null) {
-      userSession.checkGlobalPermission(UserRole.ADMIN);
-    } else {
-      if (userSession.hasGlobalPermission(GlobalPermissions.SYSTEM_ADMIN) || userSession.hasComponentUuidPermission(UserRole.ADMIN, componentUuid)) {
-        query.setComponentUuid(componentUuid);
-      } else {
-        throw new ForbiddenException("Requires administration permission");
-      }
+    String componentQuery = wsRequest.param(PARAM_COMPONENT_QUERY);
+    if (componentUuid != null && componentQuery != null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: please use https://github.com/SonarSource/sonarqube/blob/5af4d8ed1fe0ad9fc75bdf9aaf01410b66c4da5f/server/sonar-server/src/main/java/org/sonar/server/ws/WsUtils.java#L61-61

File: server/sonar-server/src/test/java/org/sonar/server/computation/ws/ActivityWsActionTest.java [hunk 5]
```diff
@@ -167,7 +177,6 @@ public void get_project_activity() {
       .setMediaType(MimeTypes.PROTOBUF)
       .execute();
 
-    // verify the protobuf response
     WsCe.ActivityResponse activityResponse = Protobuf.read(wsResponse.getInputStream(), WsCe.ActivityResponse.PARSER);
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/computation/ws/ActivityWsActionTest.java]: optional: you can use `WsCe.ActivityResponse.parseFrom`

File: server/sonar-server/src/test/java/org/sonar/server/computation/ws/ActivityWsActionTest.java [hunk 6]
```diff
@@ -175,6 +184,39 @@ public void get_project_activity() {
     assertThat(activityResponse.getTasks(0).getComponentId()).isEqualTo("PROJECT_1");
   }
 
+  @Test
+  public void search_activity_by_component_name() {
+    componentDb.insertProjectAndSnapshot(dbTester.getSession(), newProjectDto().setName("apache struts").setUuid("P1"));
+    componentDb.insertProjectAndSnapshot(dbTester.getSession(), newProjectDto().setName("apache zookeeper").setUuid("P2"));
+    componentDb.insertProjectAndSnapshot(dbTester.getSession(), newProjectDto().setName("eclipse").setUuid("P3"));
+    dbTester.getSession().commit();
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/computation/ws/ActivityWsActionTest.java]: use `dbTester.commit()`

File: sonar-db/src/main/java/org/sonar/db/ce/CeActivityDao.java [hunk 7]
```diff
@@ -64,10 +65,16 @@ public void deleteByUuid(DbSession dbSession, String uuid) {
    * Ordered by id desc -> newest to oldest
    */
   public List<CeActivityDto> selectByQuery(DbSession dbSession, CeActivityQuery query, RowBounds rowBounds) {
+    if (query.isShortCircuitedByComponentUuids()) {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/ce/CeActivityDao.java]: could be inlined with ternary operation

File: sonar-db/src/main/java/org/sonar/db/ce/CeActivityDao.java [hunk 8]
```diff
@@ -64,10 +65,16 @@ public void deleteByUuid(DbSession dbSession, String uuid) {
...
+      return Collections.emptyList();
+    }
     return mapper(dbSession).selectByQuery(query, rowBounds);
   }
 
   public int countByQuery(DbSession dbSession, CeActivityQuery query) {
+    if (query.isShortCircuitedByComponentUuids()) {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/ce/CeActivityDao.java]: could be inlined with ternary operation

File: sonar-db/src/main/java/org/sonar/db/ce/CeActivityQuery.java [hunk 9]
```diff
@@ -19,25 +19,43 @@
...
+  private List<String> componentUuids;
   private CeActivityDto.Status status;
   private String type;
   private Long minSubmittedAt;
   private Long maxExecutedAt;
 
-  @CheckForNull
-  public String getComponentUuid() {
-    return componentUuid;
+  public List<String> getComponentUuids() {
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/ce/CeActivityQuery.java]: add `@CheckForNull`

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java (hunk 1 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: I'm not sure the parameter should be generic (`q`), as it's restricted to components. Moreover description is specific to this WS.

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java (hunk 2 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: ok

File: sonar-db/src/test/java/org/sonar/db/ce/CeActivityDaoTest.java [hunk 10]
```diff
@@ -146,6 +147,24 @@ public void test_countByQuery() {
   }
 
   @Test
+  public void selectByQuery_no_results_if_shortcircuited_by_component_uuids() {
+    insert("TASK_1", REPORT, "PROJECT_1", CeActivityDto.Status.SUCCESS);
+
+    CeActivityQuery query = new CeActivityQuery();
+    query.setComponentUuids(Collections.<String>emptyList());
+    assertThat(underTest.selectByQuery(db.getSession(), query, new RowBounds(0, 10))).isEmpty();
```
Comment (teryk) [inline sonar-db/src/test/java/org/sonar/db/ce/CeActivityDaoTest.java]: I don't ask to modify anything here. But at the DAO level, we should now pass the https://github.com/SonarSource/sonarqube/blob/82564709cf06447b8a7e5fcadc609d892fc85509/sonar-plugin-api/src/main/java/org/sonar/api/utils/Paging.java#L26-26 and add a method in `DbUtils#pagingToRowBounds`

File: server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java (hunk 1 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/computation/ws/ActivityWsAction.java]: fine

**Comment by teryk** [comment]
Apply feedback and LGTM

**Pull request closed without merging**